### ISSUE TITLE
consistent segmentation fault with one batch query

### ISSUE THREAD

**Issue by nivekuil** [scylladb/scylladb#13001]
This is Scylla's bug tracker, to be used for reporting bugs only.
If you have a question about Scylla, and not a bug, please ask it in
our mailing-list at scylladb-dev@googlegroups.com or in our slack channel.

- [x] I have read the disclaimer above, and I am reporting a suspected malfunction in Scylla.

*Installation details*
Scylla version (or git commit hash): `docker.io/scylladb/scylla:5.1.5`
Cluster size: 1

`dump.zst` uploaded to 1bb343bf-450a-4f24-8645-3541efde8294

The query which always triggers it is a logged batch query with two statements, something like "DELETE FROM mytable where "table/column"=:value ... IF EXISTS" and "INSERT INTO mytable ..."

**Comment by nyh** [comment]
Do you see anything interesting in the log before the crash (assertion failure message, or something)?

**Comment by nivekuil** [comment]
nope, just 
```

**Comment by nyh** [comment]
Decoded backtrace (using http://backtrace.scylladb.com/index.html):
```
 (inlined by) std::vector<nonwrapping_interval<clustering_key_prefix>, std::allocator<nonwrapping_interval<clustering_key_prefix> > >::begin() const at /usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/bits/stl_vector.h:821
 (inlined by) std::vector<nonwrapping_interval<clustering_key_prefix>, std::allocator<nonwrapping_interval<clustering_key_prefix> > >::front() const at /usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/bits/stl_vector.h:1135
 (inlined by) cql3::statements::cas_request::find_old_row(cql3::statements::cas_row_update const&) const at ./cql3/statements/cas_request.cc:133
 (inlined by) cql3::statements::cas_request::build_cas_result_set(seastar::shared_ptr<cql3::metadata>, column_set const&, bool) const at ./cql3/statements/cas_request.cc:156
operator() at ./cql3/statements/batch_statement.cc:383
 (inlined by) seastar::shared_ptr<cql_transport::messages::result_message> std::__invoke_impl<seastar::shared_ptr<cql_transport::messages::result_message>, cql3::statements::batch_statement::execute_with_conditions(cql3::query_processor&, cql3::query_options const&, service::query_state&) const::$_11&, bool>(std::__invoke_other, cql3::statements::batch_statement::execute_with_conditions(cql3::query_processor&, cql3::query_options const&, service::query_state&) const::$_11&, bool&&) at /usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/bits/invoke.h:61
 (inlined by) std::__invoke_result<cql3::statements::batch_statement::execute_with_conditions(cql3::query_processor&, cql3::query_options const&, service::query_state&) const::$_11&, bool>::type std::__invoke<cql3::statements::batch_statement::execute_with_conditions(cql3::query_processor&, cql3::query_options const&, service::query_state&) const::$_11&, bool>(cql3::statements::batch_statement::execute_with_conditions(cql3::query_processor&, cql3::query_options const&, service::query_state&) const::$_11&, bool&&) at /usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/bits/invoke.h:96
 (inlined by) std::invoke_result<cql3::statements::batch_statement::execute_with_conditions(cql3::query_processor&, cql3::query_options const&, service::query_state&) const::$_11&, bool>::type std::invoke<cql3::statements::batch_statement::execute_with_conditions(cql3::query_processor&, cql3::query_options const&, service::query_state&) const::$_11&, bool>(cql3::statements::batch_statement::execute_with_conditions(cql3::query_processor&, cql3::query_options const&, service::query_state&) const::$_11&, bool&&) at /usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/functional:97
 (inlined by) auto seastar::internal::future_invoke<cql3::statements::batch_statement::execute_with_conditions(cql3::query_processor&, cql3::query_options const&, service::query_state&) const::$_11&, bool>(cql3::statements::batch_statement::execute_with_conditions(cql3::query_processor&, cql3::query_options const&, service::query_state&) const::$_11&, bool&&) at ././seastar/include/seastar/core/future.hh:1225
 (inlined by) operator() at ././seastar/include/seastar/core/future.hh:1596
 (inlined by) void seastar::futurize<seastar::shared_ptr<cql_transport::messages::result_message> >::satisfy_with_result_of<seastar::future<bool>::then_impl_nrvo<cql3::statements::batch_statement::execute_with_conditions(cql3::query_processor&, cql3::query_options const&, service::query_state&) const::$_11, seastar::future<seastar::shared_ptr<cql_transport::messages::result_message> > >(cql3::statements::batch_statement::execute_with_conditions(cql3::query_processor&, cql3::query_options const&, service::query_state&) const::$_11&&)::{lambda(seastar::internal::promise_base_with_type<seastar::shared_ptr<cql_transport::messages::result_message> >&&, cql3::statements::batch_statement::execute_with_conditions(cql3::query_processor&, cql3::query_options const&, service::query_state&) const::$_11&, seastar::future_state<bool>&&)#1}::operator()(seastar::internal::promise_base_with_type<seastar::shared_ptr<cql_transport::messages::result_message> >&&, cql3::statements::batch_statement::execute_with_conditions(cql3::query_processor&, cql3::query_options const&, service::query_state&) const::$_11&, seastar::future_state<bool>&&) const::{lambda()#1}>(seastar::internal::promise_base_with_type<seastar::shared_ptr<cql_transport::messages::result_message> >&&, cql3::statements::batch_statement::execute_with_conditions(cql3::query_processor&, cql3::query_options const&, service::query_state&) const::$_11&&) at ././seastar/include/seastar/core/future.hh:2136
 (inlined by) operator() at ././seastar/include/seastar/core/future.hh:1589
 (inlined by) seastar::continuation<seastar::internal::promise_base_with_type<seastar::shared_ptr<cql_transport::messages::result_message> >, cql3::statements::batch_statement::execute_with_conditions(cql3::query_processor&, cql3::query_options const&, service::query_state&) const::$_11, seastar::future<bool>::then_impl_nrvo<cql3::statements::batch_statement::execute_with_conditions(cql3::query_processor&, cql3::query_options const&, service::query_state&) const::$_11, seastar::future<seastar::shared_ptr<cql_transport::messages::result_message> > >(cql3::statements::batch_statement::execute_with_conditions(cql3::query_processor&, cql3::query_options const&, service::query_state&) const::$_11&&)::{lambda(seastar::internal::promise_base_with_type<seastar::shared_ptr<cql_transport::messages::result_message> >&&, cql3::statements::batch_statement::execute_with_conditions(cql3::query_processor&, cql3::query_options const&, service::query_state&) const::$_11&, seastar::future_state<bool>&&)#1}, bool>::run_and_dispose() at ././seastar/include/seastar/core/future.hh:781
```

@kostja @gleb-cloudius maybe you have an idea?

**Comment by nyh** [comment]
The crash happens in `cas_request::find_old_row()`  in a line that does:
```c++
const clustering_key& ckey = op.ranges.front().start() ?  op.ranges.front().start()->value() : empty_ckey;
```

If op.ranges is empty, then using op.ranges.front() can cause a crash. But I don't know what this case means. I didn't yet manage to recreate the crash. @nivekuil maybe you can share a more accurate example of a problematic query, maybe there is something I'm missing in my unsuccessful reproduction attempts.

**Comment by nivekuil** [comment]
5.2.0-rc1 no longer has the segfault, and the client returns this error from scylla: `com.datastax.oss.driver.api.core.servererrors.InvalidQueryException: Unexpected unset value for bind variable 0`. Indeed the client did not pass the named binding for the primary key in the query.

**Comment by kostja** [comment]
I believe it's a duplicate of https://github.com/scylladb/scylladb/issues/11954 (closed).

**Comment by nyh** [comment]
@kostja, according to #11954, Avi backported the fix to 5.1, and I just checked the patch "modification_statement: fix LWT insert crash if clustering key is null" is in 5.1.5 which @nivekuil said he was using. So I think it can't be exactly the same issue - but it can be something similar. @nivekuil said above that a newer Scylla says "Unexpected unset value", so while the fix for #11954 fixed the situation for a **null** value, maybe some later fix fixed it also for **unset** value (which isn't the same thing as a null)? We should backport that fix. Maybe you or @cvybhu know which fix this was.

It seems we also don't have a regression test for the unset case, I'll prepare one.

**Comment by nyh** [comment]
The following simple test in test/cql-pytest.py reproduces the bug on 5.1.5 and that it is fixed on master. I'll send a patch with the new test. @cvybhu @avikivity  we will need to consider backporting the fix, although it may be very difficult to cleanly backport (since it's probably intertwined with a lot of expression changes...) but it involves a crash...

```python
def test_unset_insert_where_lwt(cql, table2):
    p = unique_key_int()
    stmt = cql.prepare(f'INSERT INTO {table2} (p, c) VALUES ({p}, ?) IF NOT EXISTS')
    with pytest.raises(InvalidRequest, match="unset"):
        cql.execute(stmt, [UNSET_VALUE])
```

**Comment by nyh** [comment]
@mykaul please note that @kostja was right to close this issue as already-fixed (just not by the specific issue he pointed to). We do, need to consider backporting the fix. But it is likely not to be easy (@cvybhu will know better), but I don't know if the issue needs to be open for that. Maybe it just needs a backport candidate tag.

**Comment by avikivity** [comment]
@nivekuil are you using UNSET values? those are distinct from forgetting to pass a bind variable.

**Comment by nyh** [comment]
@avikivity maybe they are not really distinct (I'm surprised as you...) - I just checked and it appears the Python driver sends the same thing (and gets the same error back) for the following two statements:
```
   cql.execute(stmt, [])
   cql.execute(stmt, [UNSET_VALUE])
```

**Comment by nyh** [comment]
> have a regression test for the unset case, I'll prepare one.

Created a regression test that confirms that this bug existed in 5.1.5 but doesn't any more: https://github.com/scylladb/scylladb/pull/13007

**Comment by mykaul** [comment]
> > have a regression test for the unset case, I'll prepare one.
> 
> Created a regression test that confirms that this bug existed in 5.1.5 but doesn't any more: #13007

I've left it open exactly to evaluate if it can be fixed for 5.1.6

**Comment by nyh** [comment]
I assigned evaluating this backport to @cvybhu because I think he would be the one who'll best remember which of the latest changes fixed this bug, and how difficult it would be to either backport the same fix, or if it's a big change (I believe it is...) at least add a ad-hoc check for the UNSET in the old code.
Jan - check out the tests I wrote in https://github.com/scylladb/scylladb/pull/13007 to reproduce this issue. (you can even use them to bisect this bug, if you don't remember what fixed it).

**Comment by cvybhu** [comment]
It was most likely fixed by #12517
I ran the tests on the latest Scylla and the checks that throw the exception were added in this PR.

Sadly trying to cherry-pick the commit to branch-5.1 causes a bunch of conflicts, I will take a look how bad it is.

**Comment by cvybhu** [comment]
Posted my backport attempt in #13029

**Comment by cvybhu** [comment]
This bug was fixed in a PR that introduced a lot of changes and touched many parts of the codebase.
I don't feel comfortable backporting such a big change, especially since I'm not the original author, and Avi also thinks it's too big to backport (https://github.com/scylladb/scylladb/pull/13029#issuecomment-1448117284) .

Unless it's important to have it in 5.1 for some reasons it would be safer to skip this backport.

**Comment by cvybhu** [comment]
An alternative would be to fix this specific bug independently on `5.1`. It would most likely be a small change - adding a validation check somewhere.

**Comment by mykaul** [comment]
> An alternative would be to fix this specific bug independently on `5.1`. It would most likely be a small change - adding a validation check somewhere.

Makes more sense to me.

**Comment by cvybhu** [comment]
Okay I made #13057, which is a simplified version of #12517 for `5.1`.
It works, but introduces the same backwards incompatible changes that #12517 introduced.

**Comment by nyh** [comment]
I personally prefer option B, but don't feel strongly about it so feel free to decide on another option.

Regarding the "bonus" - I think you need to fix the code in the same place also in master. Either the empty case is perfectly fine (perhaps the case in #13129) and the code should just do nothing, OR, maybe it indicates a bug and we should on_internal_error - but in any case the dereferencing of the null pointer that we do today is wrong.

**Comment by cvybhu** [comment]
> I personally prefer option B, but don't feel strongly about it so feel free to decide on another option.

I'm also leaning a bit towards B. A is a better change, but it's bad to introduce breaking changes during a version update from `5.1.x` to `5.1.y`. I guess Scylla doesn't follow exact semver, but we probably have some rules about versioning and breaking changes (?).

**Comment by nyh** [comment]
I think we can close this issue. The original issue did not exist in master (or 5.2) because it was fixed when the UNSET handling was recently rewritten. In https://github.com/scylladb/scylladb/pull/13133 we also fixed 5.1 in a different and less intrusive way (avoiding the crash but not making other changes). I've leaving the "backport candidate" tag because we still should consider whether to (trivially?) backport the same fix of https://github.com/scylladb/scylladb/pull/13133 also to 5.0.
